agent/eni/watcher.go

```diff
@@ -0,0 +1,255 @@
+	IsMACAddressPresent(mac string) bool
+}
+
+// UdevWatcher maintains the state of attached ENIs
+// to the instance. It also has supporting elements to
+// maintain consistency and update intervals
+type UdevWatcher struct {
+	ctx                  context.Context
+	cancel               context.CancelFunc
+	updateLock           sync.RWMutex
+	updateIntervalTicker *time.Ticker
+	enis                 map[string]string // MAC => Device-Name
+	netlinkClient        netlinkWrapper.NetLink
+	udevMonitor          udevWrapper.Udev
+	events               chan *udev.UEvent
+}
+
+// New is used to return an instance of the UdevWatcher struct
+func New(ctx context.Context, nlWrap netlinkWrapper.NetLink, udevWrap udevWrapper.Udev) *UdevWatcher {
+	derivedContext, cancel := context.WithCancel(ctx)
+	return &UdevWatcher{
+		ctx:           derivedContext,
+		cancel:        cancel,
+		enis:          make(map[string]string, mapCapacityHint),
+		netlinkClient: nlWrap,
+		udevMonitor:   udevWrap,
+		events:        make(chan *udev.UEvent),
+	}
+}
+
+// Init initializes a new ENI Watcher
+func (udevWatcher *UdevWatcher) Init() error {
+	links, err := udevWatcher.netlinkClient.LinkList()
+	if err != nil {
+		return errors.Wrapf(err, "init udevWatcher: error retrieving network interfaces")
+	}
+
+	udevWatcher.updateLock.Lock()
+	for _, link := range links {
+		deviceName := link.Attrs().Name
+		macAddress := link.Attrs().HardwareAddr.String()
+		if macAddress != "" {
+			udevWatcher.addDeviceWithMACAddressUnsafe(deviceName, macAddress)
+		}
+	}
+	udevWatcher.updateLock.Unlock()
+
+	// UDev Event Handler
+	go udevWatcher.eventHandler(udevWatcher.ctx)
+	return nil
+}
+
+// Start periodically updates the state of ENIs connected to the system
+func (udevWatcher *UdevWatcher) Start() {
+	udevWatcher.performPeriodicReconciliation(udevWatcher.ctx, defaultReconciliationInterval)
+}
+
+// Stop is used to invoke the cancellation routine
+func (udevWatcher *UdevWatcher) Stop() {
+	udevWatcher.cancel()
+}
+
+// performPeriodicReconciliation is used to periodically invoke the
+// reconciliation process based on a ticker
+func (udevWatcher *UdevWatcher) performPeriodicReconciliation(ctx context.Context, updateInterval time.Duration) {
+	udevWatcher.updateIntervalTicker = time.NewTicker(updateInterval)
+	for {
+		select {
+		case <-udevWatcher.updateIntervalTicker.C:
+			udevWatcher.reconcileOnce()
+		case <-ctx.Done():
+			udevWatcher.updateIntervalTicker.Stop()
+			return
+		}
+	}
+}
+
+// reconcileOnce is used to reconcile the state of ENIs attached to the instance
+func (udevWatcher *UdevWatcher) reconcileOnce() {
+	links, err := udevWatcher.netlinkClient.LinkList()
+	if err != nil {
+		log.Warnf("Error retrieving network interfaces: %v", err)
+		return
+	}
+
+	currentState := udevWatcher.buildState(links)
+
+	// Return on empty state
+	if len(currentState) == 0 {
+		log.Info("State built for reconciliation is empty!")
```

Comment:
Please update this comment to have more details about where it's originating from: `"Udev watcher: State built..."`